File: ui-tests/PageObjects/Components/ValueComponentView.cs
using System;
using System.Threading.Tasks;
using Microsoft.Playwright;

namespace UiTests.PageObjects.Components;

/// <summary>
/// Lightweight wrapper around the reusable value component markup used across panes,
/// including scratchpad entries, tooltips, and inline state values.
/// </summary>
public class ValueComponentView
{
    private readonly ILocator _root;

    public ValueComponentView(ILocator root)
    {
        _root = root ?? throw new ArgumentNullException(nameof(root));
    }

    /// <summary>
    /// Root locator representing the <c>.value-expanded</c> element.
    /// </summary>
    public ILocator Root => _root;

    private ILocator NameContainer() => _root.Locator(".value-name-container");
    private ILocator ExpandButton() => NameContainer().Locator(".value-expand-button");
    private ILocator AddToScratchpadButton() => NameContainer().Locator(".add-to-scratchpad-button");
    private ILocator HistoryToggle() => NameContainer().Locator(".toggle-value-history");
    private ILocator ValueTypeNode() => NameContainer().Locator(".value-type");
    private ILocator ValueTextNode() => _root.Locator(".value-expanded-text").First;

    /// <summary>
    /// Extracts the label rendered before the value (e.g. <c>remaining_shield:</c>).
    /// </summary>
    public async Task<string> NameAsync()
    {
        var nameText = await NameContainer().Locator(".value-name").InnerTextAsync();
        var name = (nameText ?? string.Empty).Trim();
        return name.TrimEnd(':').Trim();
    }

    /// <summary>
    /// Returns the textual representation of the value body.
    /// </summary>
    public async Task<string> ValueTextAsync()
    {
        var valueText = await ValueTextNode().InnerTextAsync();
        return (valueText ?? string.Empty).Trim();
    }

    /// <summary>
    /// Returns the type annotation if the component exposes one.
    /// </summary>
    public async Task<string?> ValueTypeAsync()
    {
        if (await ValueTypeNode().CountAsync() == 0)
        {
            return null;
        }

        var text = await ValueTypeNode().First.InnerTextAsync();
        return text?.Trim();
    }

    /// <summary>
    /// Indicates whether the value supports nested expansion.
    /// </summary>
    public async Task<bool> IsExpandableAsync()
        => await ExpandButton().Locator(".caret-expand, .caret-collapse").CountAsync() > 0;

    /// <summary>
    /// Attempts to expand the value component (no-op if it is not expandable).
    /// </summary>
    public async Task ExpandAsync()
    {
        if (await IsExpandableAsync())
        {
            await ExpandButton().ClickAsync();
        }
    }

    /// <summary>
    /// Determines whether the inline "Add to scratchpad" button is available.
    /// </summary>
    public async Task<bool> HasAddToScratchpadButtonAsync()
        => await AddToScratchpadButton().CountAsync() > 0;

    /// <summary>
    /// Clicks the inline "Add to scratchpad" button (throws if it is absent).
    /// </summary>
    public async Task ClickAddToScratchpadAsync()
    {
        if (!await HasAddToScratchpadButtonAsync())
        {
            throw new InvalidOperationException("This value component does not expose an inline Add to scratchpad button.");
        }

        await AddToScratchpadButton().ClickAsync();
    }

    /// <summary>
    /// Indicates whether the value component offers a history toggle entry.
    /// </summary>
    public async Task<bool> HasHistoryToggleAsync()
        => await HistoryToggle().CountAsync() > 0;

    /// <summary>
    /// Invokes the history toggle (throws when unavailable).
    /// </summary>
    public async Task ToggleHistoryAsync()
    {
        if (!await HasHistoryToggleAsync())
        {
            throw new InvalidOperationException("No history toggle is available for this value component.");
        }

        await HistoryToggle().ClickAsync();
    }
}

File: src/tests/gui/page-objects/components/value-component-view.ts
import type { Locator } from "@playwright/test";

/**
 * Lightweight wrapper around the reusable value component markup used across
 * panes, including scratchpad entries, tooltips, and inline state values.
 *
 * Port of ui-tests/PageObjects/Components/ValueComponentView.cs
 */
export class ValueComponentView {
  readonly root: Locator;

  constructor(root: Locator) {
    this.root = root;
  }

  private nameContainer(): Locator {
    return this.root.locator(".value-name-container");
  }

  private expandButton(): Locator {
    return this.nameContainer().locator(".value-expand-button");
  }

  private addToScratchpadButton(): Locator {
    return this.nameContainer().locator(".add-to-scratchpad-button");
  }

  /** toggle-value-history is a sibling of value-name-container, not inside it */
  private historyToggle(): Locator {
    return this.root.locator(".toggle-value-history");
  }

  private valueTypeNode(): Locator {
    return this.root.locator(".value-type");
  }

  private valueTextNode(): Locator {
    return this.root.locator(".value-expanded-text").first();
  }

  /**
   * Extracts the label rendered before the value (e.g. `remaining_shield:`).
   */
  async name(): Promise<string> {
    const nameText = await this.nameContainer().locator(".value-name").innerText();
    return nameText.trim().replace(/:$/, "").trim();
  }

  /**
   * Returns the textual representation of the value body.
   */
  async valueText(): Promise<string> {
    return (await this.valueTextNode().innerText()).trim();
  }

  /**
   * Returns the type annotation if the component exposes one.
   */
  async valueType(): Promise<string | null> {
    if ((await this.valueTypeNode().count()) === 0) {
      return null;
    }
    return (await this.valueTypeNode().first().innerText()).trim();
  }

  /**
   * Indicates whether the value supports nested expansion.
   */
  async isExpandable(): Promise<boolean> {
    return (
      (await this.expandButton().locator(".caret-expand, .caret-collapse").count()) > 0
    );
  }

  /**
   * Attempts to expand the value component (no-op if not expandable).
   */
  async expand(): Promise<void> {
    if (await this.isExpandable()) {
      await this.expandButton().click();
    }
  }

  /**
   * Determines whether the inline "Add to scratchpad" button is available.
   */
  async hasAddToScratchpadButton(): Promise<boolean> {
    return (await this.addToScratchpadButton().count()) > 0;
  }

  /**
   * Clicks the inline "Add to scratchpad" button.
   */
  async clickAddToScratchpad(): Promise<void> {
    if (!(await this.hasAddToScratchpadButton())) {
      throw new Error(
        "This value component does not expose an inline Add to scratchpad button.",
      );
    }
    await this.addToScratchpadButton().click();
  }

  /**
   * Indicates whether the value component offers a history toggle.
   */
  async hasHistoryToggle(): Promise<boolean> {
    return (await this.historyToggle().count()) > 0;
  }

  /**
   * Invokes the history toggle.
   */
  async toggleHistory(): Promise<void> {
    if (!(await this.hasHistoryToggle())) {
      throw new Error(
        "No history toggle is available for this value component.",
      );
    }
    await this.historyToggle().click();
  }
}
